Revise queue splitting docs (wip) (#22)
metalbear-co/docs: using-mirrord/queue-splitting.md

# Queue Splitting

> If your application consumes messages from a message broker (e.g. Kafka cluster), you should choose a configuration that matches your intention:

Reviewer: More confusing this way IMO

> 1. If you're ok with your local application competing for queue messages with the remote target, and with your teammates' mirrord sessions — run the application with mirrord without any special configuration.

Reviewer: Why the rephrase? Sounds kind of passive aggressive instead of just aggressive

2. If you want your local application to be an exclusive consumer of queue messages — run the application with [`copy_target` + `scale_down`](copy-target.md#replacing-a-whole-deployment-using-scale_down) features.
3. If you want to precisely control which messages will be consumed by your local application — run the application with the queue splitting feature. The allows you define a message filter in your mirrord configuration. All messages matching that filter will be redirected by the mirrord operator to your local application. Other messages will **not** reach your local application. 

{% hint style="info" %}

> This feature is only relevant for users on the Team and Enterprise pricing plans.

Reviewer: ```suggestion
This feature is only available for users on the Team and Enterprise pricing plans.
```

{% endhint %}

{% hint style="info" %}

> So far queue splitting is available for [Amazon SQS](https://aws.amazon.com/sqs/) and [Kafka](https://kafka.apache.org/). Pretty soon we'll support RabbitMQ as well.

Reviewer: ```suggestion
Queue Splitting is currently available for [Amazon SQS](https://aws.amazon.com/sqs/) and [Kafka](https://kafka.apache.org/). Pretty soon we'll support RabbitMQ as well.
```

{% endhint %}

## How It Works

> When a Kafka splitting session starts, the mirrord operator patches the target workload (e.g. deployment or rollout) to consume messages from a different, temporary queue or topic.

Reviewer: Why are you talking specifically about Kafka here?

Author: oopsie

> That temporary queue/topic is *exclusive* to the target workload, and its name is randomized.

Reviewer: Confusing way to tell the story here - shouldn't you first mention that the operator creates these new temporary queues/topics, and sets itself as the consumer of the original queue/topic?

> That temporary queue/topic is *exclusive* to the target workload, and its name is randomized.

Author: Oke, it's not

> Similarly, the local application is redirected to consume messages from its own *exclusive* temporary queue or topic.

Reviewer: ```suggestion
Similarly, the local application is reconfigured to consume messages from its own *exclusive* temporary queue or topic.
```

When the first mirrord Kafka splitting session starts, two temporary topics are created (one for the target deployed in the cluster, one for the user's local application),
and the mirrord operator routes messages according to the [user's filter](queue-splitting.md#setting-a-filter-for-a-mirrord-run):

![One Kafka splitting session](queue-splitting/1-user-kafka.svg)

If a second user then starts a mirrord Kafka splitting session on the same topic, a third temporary topic is created (for the second user's local application).
The mirrord operator includes the new topic and the second user's filter in the routing logic.

![Two Kafka splitting sessions](queue-splitting/2-users-kafka.svg)

If the filters defined by the two users both match some message, one of the users will receive the message at random.

{% endtab %}

{% endtabs %}

Temporary queues and topics are managed by the mirrord operator and garbage collected in the background. After all queue splitting sessions end, the operator promptly deletes the allocated resources.

Plese note that:
1. Temporary queues and topics created for the deployed targets will not be deleted as long as there are any targets' pods that use them.

> 2. In case of SQS splitting, deployed targets will remain redirected as long as their temporary queues have unconsumed messages.

Author: Changed.

> Also why is this SQS only?

Because Kafka splitting does not have this :shrug:
